src/clickonce/launcher/HostFinder.cs

```diff
@@ -33,23 +38,59 @@ internal static string GetHost(string applicationFilePath)
         {
             string arch = GetProcessorArchitectureFromAssembly(applicationFilePath);
 
-            if (arch == "x86")
+            if (arch == ArchX86)
             {
-                return GetX86Host();
+                return GetGlobalHost(arch);
             }
-            else if (arch == "amd64" || arch == "arm64")
+            else if (arch == ArchX64 || arch == ArchArm64)
             {
-                return Get64bitHost();
+                return Environment.Is64BitOperatingSystem ? GetGlobalHost(arch) : string.Empty;
             }
             else if (arch == "msil")
             {
-                string host = Get64bitHost();
-                return string.IsNullOrEmpty(host) ? GetX86Host() : host;
+                string host = string.Empty;
+
+                // On arm64 systems, first look for arm64 host
+                if (IsArm64System)
+                {
+                    host = GetGlobalHost(ArchArm64);
+                }
+
+                // Fallback for arm64 systems, and native x64 systems
+                if (string.IsNullOrEmpty(host))
+                {
+                    host = GetGlobalHost(ArchX64);
+                }
+
+                return string.IsNullOrEmpty(host) ? GetGlobalHost(ArchX86) : host;
             }
 
             return string.Empty;
         }
 
+        /// <summary>
+        /// Checks if running on Arm64 system
+        /// </summary>
+        private static bool IsArm64System
+        {
+            get
+            {
+                string proc = Environment.GetEnvironmentVariable("PROCESSOR_ARCHITECTURE");
```

Comment:
Can you use System.Runtime.InteropServices.RuntimeInformation.ProcessArchitecture instead of the environment variable?

Reply:
> Can you use System.Runtime.InteropServices.RuntimeInformation.ProcessArchitecture instead of the environment variable?

It seems that the API returns architecture of the currently running app: https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.processarchitecture?view=net-6.0&viewFallbackFrom=netframework-4.6.2#system-runtime-interopservices-runtimeinformation-processarchitecture

For ClickOnce deployment, this is not guaranteed to match the OS architecture, even though Launcher is built as MSIL.

Reply:
For OS Architecture, there is the RuntimeInformation.OSArchitecture property.

Reply:
> For OS Architecture, there is the RuntimeInformation.OSArchitecture property.

That property is available in v4.7.1 and later. Launcher targets v4.5, to ensure we can run on as many systems as possible.

Reply:
I think the following code can be used as a model, or simply copied:

https://github.com/microsoft/vstest/blob/fcfbccd86e90bc02a4e59386e725aa41377a3b92/src/Microsoft.TestPlatform.PlatformAbstractions/net451/System/PlatformEnvironment.cs#L32-L56

Reply:
Actually, the following might be needed: https://github.com/dotnet/runtime/blob/cb5f0afd7fa056ee720864985da424090183d7d8/src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/RuntimeInformation.Windows.cs#L34-L74

Reply:
I've pushed the commit that fixes this.